examples/tcp-wait/server.rs

```diff
@@ -0,0 +1,236 @@
+        let qts_cancelled: Vec<QToken> = self.cancel_pending_operations(qd);
+        self.unregister_client(qd);
+        self.issue_close(qd)?;
+        self.clients_closed += 1;
+        println!("{} clients closed", self.clients_closed);
+        Ok(qts_cancelled)
+    }
+
+    // Attempt to wait for a push() operation to complete after asynchronous closing a socket.
+    pub fn run(&mut self) -> Result<()> {
+        self.libos.listen(self.sockqd, self.nclients)?;
+        self.issue_accept()?;
+
+        loop {
+            // Stop when enough connections have been terminated.
+            if self.clients_closed >= self.nclients {
+                // Sanity check that all connections have been closed.
+                assert_eq!(
+                    self.clients.len(),
+                    0,
+                    "there should be no clients connected, but there are"
+                );
+                break;
+            }
+
+            let qr: demi_qresult_t = {
+                // If error, close sockets.
+                // FIXME: https://github.com/demikernel/demikernel/issues/640
+                let (index, qr): (usize, demi_qresult_t) = self.libos.wait_any(&self.qts, None)?;
+                self.mark_completed_operation(index)?;
+                qr
+            };
+
+            match qr.qr_opcode {
+                // Accept completed.
+                demi_opcode_t::DEMI_OPC_ACCEPT => {
+                    let qd: QDesc = unsafe { qr.qr_value.ares.qd.into() };
+
+                    // Handles the completion of an accept() operation.
+                    self.handle_connection_establishment(qd)?;
+
+                    // Accept more connections.
+                    self.issue_accept()?;
+                },
+                // Pop completed.
+                demi_opcode_t::DEMI_OPC_POP => {
+                    let qd: QDesc = qr.qr_qd.into();
+                    let sga: demi_sgarray_t = unsafe { qr.qr_value.sga };
+
+                    self.libos.sgafree(sga)?;
+
+                    // Handle connection termination.
+                    let qts_cancelled: Vec<QToken> = self.handle_connection_termination(qd)?;
+
+                    // Ensure that the client has no pending operations.
+                    assert!(
+                        qts_cancelled.is_empty(),
+                        "client should not have any pending operations, but it has"
+                    );
+                },
+                demi_opcode_t::DEMI_OPC_FAILED => {
+                    let qd: QDesc = qr.qr_qd.into();
+                    let errno: i32 = qr.qr_ret;
+
+                    // Ensure that this error was triggered because
+                    // the client has terminated the connection.
+                    assert_eq!(
+                        errno,
+                        libc::ECONNRESET,
+                        "client should have had terminated the connection, but it has not"
+                    );
```

Comment:
I would add this condition check in the matching rule, because it reads cleaner and you would panic anyways:

`demi_opcode_t::DEMI_OPC_FAILED if qr.qr_ret == libc::ECONNRESET`

Reply:
Done